x/lightclient/ante/ibc_msg_update_client.go

```diff
@@ -1,82 +1,153 @@
+	return errorsmod.Wrap(i.k.SaveSigner(ctx, seq.Address, msg.ClientId, h), "save updater")
+}
+
+var (
+	errIsMisbehaviour   = errorsmod.Wrap(gerrc.ErrFailedPrecondition, "misbehavior evidence is disabled for canonical clients")
+	errNoHeader         = errors.New("message does not contain header")
+	errProposerMismatch = errorsmod.Wrap(gerrc.ErrInvalidArgument, "validator set proposer not equal header proposer field")
+)
+
+func (i IBCMessagesDecorator) getSequencer(ctx sdk.Context, header *ibctm.Header) (sequencertypes.Sequencer, error) {
+	proposerBySignature := header.ValidatorSet.Proposer.GetAddress()
+	proposerByData := header.Header.ProposerAddress
+	// Does ibc already guarantee this equal to header.ProposerAddr? I don't think so
+	if !bytes.Equal(proposerBySignature, proposerByData) {
+		return sequencertypes.Sequencer{}, errProposerMismatch
 	}
+	return i.k.SeqK.SequencerByDymintAddr(ctx, proposerByData)
+}
 
+func getHeader(msg *ibcclienttypes.MsgUpdateClient) (*ibctm.Header, error) {
 	clientMessage, err := ibcclienttypes.UnpackClientMessage(msg.ClientMessage)
 	if err != nil {
-		return nil
+		return nil, errorsmod.Wrap(err, "unpack client message")
 	}
-	_, ok = clientMessage.(*ibctm.Misbehaviour)
+	_, ok := clientMessage.(*ibctm.Misbehaviour)
 	if ok {
-		return errorsmod.Wrap(gerrc.ErrFailedPrecondition, "misbehavior evidence is disabled for canonical clients")
+		return nil, errIsMisbehaviour
 	}
 	header, ok := clientMessage.(*ibctm.Header)
 	if !ok {
-		return nil
+		return nil, errNoHeader
 	}
+	return header, nil
+}
+
+// if containingHPlus1 is not nil then containingH also guaranteed to not be nil
+type stateInfos struct {
+	containingH      *rollapptypes.StateInfo
+	containingHPlus1 *rollapptypes.StateInfo
+}
 
+// getStateInfos gets state infos for h and h+1
+func (i IBCMessagesDecorator) getStateInfos(ctx sdk.Context, rollapp string, h uint64) (stateInfos, error) {
 	// Check if there are existing block descriptors for the given height of client state
-	height := uint64(header.Header.Height)
-	stateInfo, err := i.rollappKeeper.FindStateInfoByHeight(ctx, rollappID, height)
-	if err != nil {
-		// No BDs found for given height.
-		// Will accept the update optimistically
-		// But also save the blockProposer address with the height for future verification
-		i.acceptUpdateOptimistically(ctx, msg.ClientId, header)
-		return nil
+	s0, err := i.raK.FindStateInfoByHeight(ctx, rollapp, h)
+	if errorsmod.IsOf(err, gerrc.ErrNotFound) {
+		return stateInfos{}, nil
 	}
-	bd, _ := stateInfo.GetBlockDescriptor(height)
-
-	stateInfo, err = i.rollappKeeper.FindStateInfoByHeight(ctx, rollappID, height+1)
 	if err != nil {
-		// No BDs found for next height.
-		// Will accept the update optimistically
-		// But also save the blockProposer address with the height for future verification
-		i.acceptUpdateOptimistically(ctx, msg.ClientId, header)
-		return nil
+		return stateInfos{}, err
+	}
+	s1 := s0
+	if !s1.ContainsHeight(h) {
```

Comment:
how it's possbile?
if h not found in `s0`, it means it's future height
how `h+1` can be found?
maybe u meant `if !s0.ContainsHeight(h+1) {`?

Reply:
great catch! Thanks